salt/utils/subprocess.py

```diff
@@ -4,12 +4,50 @@
 '''
 from __future__ import absolute_import, print_function, unicode_literals
 
+import os
+import sys
 import shlex
+import logging
 import subprocess
 import threading
 import salt.exceptions
-import salt.utils.data
 from salt.ext import six
+from salt.ext.six.moves import range
+
+
+log = logging.getLogger(__name__)
+
+
+class FdPopen(subprocess.Popen):
+    '''
+    Python2 closing file descriptors in a blind way by a system's given range.
+    Therefore close_fds can significantly slow down the system in various circumstances.
+    '''
+
+    if sys.platform.startswith('linux') and six.PY2:
+        def _close_fds(self, but):
+            '''
+            Close only file descriptors that are needed.
+            The issue is not new: https://bugs.python.org/issue1663329
+
+            Do *not* use os.closerange (!)
+
+            :param but:
+            :return:
+            '''
+            proc_path = '/proc/{}/fd'.format(os.getpid())
```

Comment:
Check if 
FreeBSD, Darwin or OS X check /dev/fd/
else if exist dir /proc/self/fd 
else if exist dir  dir /proc/<pid>/fd

/proc/pid/fd/ or /proc/self/fd/ (Linux, Solaris, AIX, Cygwin, NetBSD)
(AIX does not support "self")

/dev/fd/ (FreeBSD, Darwin, OS X)

Reply:
@twangboy what is the os_family for FreeBSD, Darwin or OS X

Reply:
@damon-atkins while thanks for the valuable info about other OSes, unfortunately I am not going to support them in this PR as I have none of them around. However, feel free to do another PR on top of this, which will do the following:

- Add another check for the OS platform
- Get an appropriate path of FDs
- Include unit tests for that

I am happy to assist you in your PR with any questions you will have.